OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyMapPreviewLogic.cs

```diff
@@ -22,100 +24,39 @@ public class LobbyMapPreviewLogic : ChromeLogic
 		int blinkTick;
 
 		[ObjectCreator.UseCtor]
-		internal LobbyMapPreviewLogic(Widget widget, ModData modData, OrderManager orderManager, LobbyLogic lobby)
+		internal LobbyMapPreviewLogic(Widget widget, ModData modData, OrderManager orderManager, Func<MapPreview> getMap,
+			Action<MapPreviewWidget, MapPreview, MouseInput> onMouseDown, Func<MapPreview, Dictionary<CPos, SpawnOccupant>> getSpawnOccupants)
 		{
 			var mapRepository = modData.Manifest.Get<WebServices>().MapRepository;
 
 			var available = widget.GetOrNull("MAP_AVAILABLE");
 			if (available != null)
 			{
-				available.IsVisible = () => lobby.Map.Status == MapStatus.Available && (!lobby.Map.RulesLoaded || !lobby.Map.InvalidCustomRules);
-
-				var preview = available.Get<MapPreviewWidget>("MAP_PREVIEW");
-				preview.Preview = () => lobby.Map;
-				preview.OnMouseDown = mi => LobbyUtils.SelectSpawnPoint(orderManager, preview, lobby.Map, mi);
-				preview.SpawnOccupants = () => LobbyUtils.GetSpawnOccupants(orderManager.LobbyInfo, lobby.Map);
-
-				var titleLabel = available.GetOrNull<LabelWithTooltipWidget>("MAP_TITLE");
-				if (titleLabel != null)
-				{
-					var font = Game.Renderer.Fonts[titleLabel.Font];
-					var title = new CachedTransform<MapPreview, string>(m => WidgetUtils.TruncateText(m.Title, titleLabel.Bounds.Width, font));
-					titleLabel.GetText = () => title.Update(lobby.Map);
-					titleLabel.GetTooltipText = () => lobby.Map.Title;
-				}
-
-				var typeLabel = available.GetOrNull<LabelWidget>("MAP_TYPE");
-				if (typeLabel != null)
-				{
-					var type = new CachedTransform<MapPreview, string>(m => lobby.Map.Categories.FirstOrDefault() ?? "");
-					typeLabel.GetText = () => type.Update(lobby.Map);
-				}
-
-				var authorLabel = available.GetOrNull<LabelWidget>("MAP_AUTHOR");
-				if (authorLabel != null)
-				{
-					var font = Game.Renderer.Fonts[authorLabel.Font];
-					var author = new CachedTransform<MapPreview, string>(
-						m => WidgetUtils.TruncateText("Created by {0}".F(lobby.Map.Author), authorLabel.Bounds.Width, font));
-					authorLabel.GetText = () => author.Update(lobby.Map);
-				}
+				available.IsVisible = () => getMap().Status == MapStatus.Available && (!getMap().RulesLoaded || !getMap().InvalidCustomRules);
```

Comment:
The natural question that follows this is whether we can cache `getMap` across the different widget bindings.  The answer is probably, but it is probably not worth it.  You could in principle override the `ChromeLogic.Tick` and then set a field in the class that is used in all the bindings, but I expect that this would introduce a noticable latency, and possibly other side effects.

So long as the map usage is consistent within individual bindings, it's generally not a big deal if it changes between evaluating different widgets.  The worst case scenario is that e.g. the title and author labels are out of sync for 1/60th of a second until the next frame is rendered.